examples/mnist_lightning.py

```diff
@@ -0,0 +1,169 @@
+
+To start training:
+$ python mnist_lightning.py fit
+
+More information about setting training parameters:
+$ python mnist_lightning.py fit --help
+
+To see logs:
+$ tensorboard --logdir=lightning_logs/
+"""
+
+import os
+import warnings
+
+import pytorch_lightning as pl
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+import torchmetrics
+from opacus import PrivacyEngine
+from opacus.data_loader import DPDataLoader
+from opacus.lightning import DPLightningDataModule
+from pl_bolts.datamodules import MNISTDataModule
+from pytorch_lightning.utilities.cli import LightningCLI
+
+warnings.filterwarnings("ignore")
+
+
+class LitSampleConvNetClassifier(pl.LightningModule):
+    def __init__(
+        self,
+        lr: float = 0.1,
+        enable_dp: bool = True,
+        delta: float = 1e-5,
+        noise_multiplier: float = 1.0,
+        max_grad_norm: float = 1.0,
+    ):
+        """A simple conv-net for classifying MNIST with differential privacy training
+
+        Args:
+            lr: Learning rate
+            enable_dp: Enables training with privacy guarantees using Opacus (if True), vanilla SGD otherwise
+            delta: Target delta for which (eps, delta)-DP is computed
+            noise_multiplier: Noise multiplier
+            max_grad_norm: Clip per-sample gradients to this norm
+        """
+        super().__init__()
+
+        # Hyper-parameters
+        self.lr = lr
+
+        # Parameters
+        self.conv1 = nn.Conv2d(1, 16, 8, 2, padding=3)
+        self.conv2 = nn.Conv2d(16, 32, 4, 2)
+        self.fc1 = nn.Linear(32 * 4 * 4, 32)
+        self.fc2 = nn.Linear(32, 10)
+
+        # Metrics
+        self.test_accuracy = torchmetrics.Accuracy()
+
+        # Differential privacy
+        self.enable_dp = enable_dp
+        self.delta = delta
+        self.noise_multiplier = noise_multiplier
+        self.max_grad_norm = max_grad_norm
+        if self.enable_dp:
+            self.privacy_engine = PrivacyEngine()
+
+    def forward(self, x):
+        # x of shape [B, 1, 28, 28]
+        x = F.relu(self.conv1(x))  # -> [B, 16, 14, 14]
+        x = F.max_pool2d(x, 2, 1)  # -> [B, 16, 13, 13]
+        x = F.relu(self.conv2(x))  # -> [B, 32, 5, 5]
+        x = F.max_pool2d(x, 2, 1)  # -> [B, 32, 4, 4]
+        x = x.view(-1, 32 * 4 * 4)  # -> [B, 512]
+        x = F.relu(self.fc1(x))  # -> [B, 32]
+        x = self.fc2(x)  # -> [B, 10]
+        return x
+
+    def configure_optimizers(self):
+        optimizer = optim.SGD(self.parameters(), lr=self.lr, momentum=0)
+
+        if self.enable_dp:
+            # dirty introspection of the trainer instance to get the training data
+            data_loader = (
+                self.trainer._data_connector._train_dataloader_source.dataloader()
+            )
+            # transform (model, optimizer, dataloader) to DP-versions
+            model, optimizer, dataloader = self.privacy_engine.make_private(
```

Comment:
Just trying to understand what's happening here:
data_loader on line 91 is already a DPDataLoader, right?
So we don't care that data_loader returned from `make_private` is discarded

Reply:
Following the Lightning philosophy, the module should be independent from data (we may pass different data to `trainer.fit(module, data)`). This means here we shouldn't expect wrapped data_loader automatically

Here `make_private` does nothing with the dataloader because we pass `DPLightningDataModule` to `trainer.fit`. 

But it we do not pass wrapped model, this works incorrectly (we will use the original dataloader). I'm going to add a check here and raise if the data is not wrapped and we do poisson sampling, or if we cannot get sample_rate.